pkg/core/klinebuilder.go

```diff
@@ -0,0 +1,207 @@
+//   - symbol: symbol to trace on
+//   - minInterval: unit interval, related to your signal timeframe.
+//     All the supported intervals of the binding stream should be multiple of this interval.
+func NewKLineBuilder(symbol string, minInterval types.Interval, isBackTesting bool) *kLineBuilder {
+	logger := logrus.WithField("symbol", symbol)
+	return &kLineBuilder{
+		symbol:        symbol,
+		minInterval:   minInterval,
+		intervals:     make(map[types.Interval]struct{}),
+		isBackTesting: isBackTesting,
+		klinesMap:     make(map[types.Interval]*types.KLine),
+		logger:        logger,
+	}
+}
+
+func (c *kLineBuilder) Subscribe(interval types.Interval) {
+	c.intervals[interval] = struct{}{}
+}
+
+func (c *kLineBuilder) BindStream(stream types.Stream) {
+	if c.isBackTesting {
+		stream.OnKLineClosed(c.handleKLineClosed)
+		stream.OnKLine(c.handleKLine)
+		return
+	}
+	stream.OnMarketTrade(c.handleMarketTrade)
+}
+
+func (c *kLineBuilder) handleKLine(kline types.KLine) {
+	c.EmitKLine(kline)
+}
+
+func (c *kLineBuilder) handleKLineClosed(kline types.KLine) {
+	c.EmitKLineClosed(kline)
+}
+
+func (c *kLineBuilder) handleMarketTrade(trade types.Trade) {
+	// the market data stream may subscribe to trades of multiple symbols
+	// so we need to check if the trade is for the symbol we are interested in
+	if trade.Symbol != c.symbol {
+		return
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.price = trade.Price
+	c.close = trade.Price
+	if !c.high.IsZero() {
+		if c.price.Compare(c.high) > 0 {
+			c.high = c.price
+		}
+	} else {
+		c.high = c.price
+	}
+	if !c.low.IsZero() {
+		if c.price.Compare(c.low) < 0 {
+			c.low = c.price
+		}
+	} else {
+		c.low = c.price
+	}
+	if c.open.IsZero() {
+		c.open = c.price
+	}
+	c.volume = c.volume.Add(trade.Quantity)
+	c.quoteVolume = c.quoteVolume.Add(trade.QuoteQuantity)
+	c.numTrades++
+	c.updateTime = trade.Time
+}
+
+func (c *kLineBuilder) Run(ctx context.Context) {
+	if len(c.intervals) == 0 || c.isBackTesting {
+		return
+	}
+	minDuration := c.minInterval.Duration()
+	for interval := range c.intervals {
+		if interval.Duration()%minDuration != 0 {
+			err := fmt.Errorf("interval %s is not a multiple of minInterval %s", interval, c.minInterval)
+			c.logger.Error(err)
+			panic(err)
+		}
+	}
+	c.logger.Infof("kline updater started for %s (%+v)", c.symbol, c.intervals)
+	go c.update(ctx)
+}
+
+func (c *kLineBuilder) update(ctx context.Context) {
+	// wait for the next interval
+	startTime := time.Now()
+	c.updateTime = types.Time(startTime) // kick-start update time
+	startTimeRound := startTime.Round(c.minInterval.Duration())
+	waitDuration := startTimeRound.Sub(startTime)
+	if waitDuration > 0 {
+		select {
+		case <-time.After(waitDuration):
+		case <-ctx.Done():
+			return
+		}
+	}
+	// start ticker
+	intervalDuration := c.minInterval.Duration()
+	ticker := time.NewTicker(intervalDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case tickTime := <-ticker.C:
```

Comment:
is ticker frequency 1s?

Reply:
No, it will be the minimum of supported interval on Coinbase.
See `init()` please.